geode-core/src/main/java/org/apache/geode/cache/query/security/GeodeBasedMethodAuthorizer.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.geode.cache.query.security;
+
+import java.lang.reflect.Method;
+import java.util.Objects;
+import java.util.Properties;
+
+import org.apache.geode.cache.Cache;
+import org.apache.geode.cache.Declarable;
+import org.apache.geode.cache.Region;
+
+/**
+ * An immutable and thread-safe {@link MethodInvocationAuthorizer} that allows any method execution
+ * as long as the target object does not belong to a Geode package, or does belong but it's marked
+ * as
+ * safe (see {@link RestrictedMethodAuthorizer#isAllowedGeodeMethod(Method, Object)}).
+ * <p/>
+ *
+ * Some known dangerous methods (like {@link Object#getClass()}) are also rejected by this
+ * {@link MethodInvocationAuthorizer} implementation, no matter whether the target object belongs
+ * to Geode or not (see {@link RestrictedMethodAuthorizer#isKnownDangerousMethod(Method, Object)}).
+ * <p/>
+ *
+ * This {@link MethodInvocationAuthorizer} implementation addresses only three of the four known
+ * security risks: {@code Java Reflection}, {@code Cache Modification} and
+ * {@code Region Modification}.
+ * <p/>
+ *
+ * The {@code Region Entry Modification} security risk still exists: users with the
+ * {@code DATA:READ:RegionName} privilege will be able to execute ANY method (even mutators) on the
+ * objects stored within the region and on instances used as bind parameters of the OQL, so this
+ * {@link MethodInvocationAuthorizer} implementation must be used with extreme care.
+ * <p/>
+ *
+ * Usage of this {@link MethodInvocationAuthorizer} implementation is only recommended for secured
+ * clusters on which only trusted users and applications have access to the OQL engine. It might
+ * also
+ * be used on clusters on which the entries stored are immutable.
+ * <p/>
+ *
+ * @see org.apache.geode.cache.Cache
+ * @see org.apache.geode.cache.query.security.MethodInvocationAuthorizer
+ * @see org.apache.geode.cache.query.security.RestrictedMethodAuthorizer
+ */
+public final class GeodeBasedMethodAuthorizer implements MethodInvocationAuthorizer {
```

Comment:
The name "GeodeBased..." seems vague. Could you consider a more descriptive name for this class? Maybe: RestrictUnsafeGeodeMethods

Reply:
The names were discussed and approved during the original RFC so, unless we come up with something that's extremely more meaningful, I'm resilient to change this one just yet.
What about `GeodeAllowedMethodAuthorizer`, sounds better?, `GeodePackageSecuredMethodAuthorizer`?, `GeodePackageBasedMethodAuthorizer`?.